contributors/design-proposals/node/limit-node-object-self-control.md

```diff
@@ -0,0 +1,127 @@
+
+Since the nodes self reports labels upon registration, an intruder can easily
+register a compromised node with label `foo/dedicated=customer-info-app`. The
+scheduler will then bind `customer-info-app` to the compromised node potentially
+giving the intruder easy access to the PII.
+
+This attack also extends to secrets. Suppose company `foo` runs their outward
+facing nginx on dedicated nodes to reduce exposure to the company's publicly
+trusted server certificates. They use the secret mechanism to distribute the
+serving certificate key. An intruder captures the dedicated nginx workload in
+the same way and can now use the node certificate to read the company's serving
+certificate key.
+
+## Proposed Solution
+
+In many environments, we can improve the situation by centralizing reporting of
+sensitive node attributes to a more trusted source and disallowing reporting of
+these attributes from the kubelet.
+
+### Label And Taint Restriction
+
+An operator will configure a whitelist of taints and labels that nodes are
+allowed to set on themselves. This list should include the taints and labels
+that the kubelet is already setting on itself.
+
+Well known taint keys:
+```
+node.kubernetes.io/not-ready
+node.alpha.kubernetes.io/notReady
+node.kubernetes.io/unschedulable
+node.kubernetes.io/out-of-disk
+node.kubernetes.io/memory-pressure
+node.kubernetes.io/disk-pressure
+node.kubernetes.io/network-unavailable
+node.kubernetes.io/pid-pressure
+node.cloudprovider.kubernetes.io/uninitialized
+```
+
+Well known label keys:
+
+```
+kubernetes.io/hostname
+failure-domain.beta.kubernetes.io/zone
+failure-domain.beta.kubernetes.io/region
+beta.kubernetes.io/instance-type
+beta.kubernetes.io/os
+beta.kubernetes.io/arch
+```
+
+As well as any taints and labels that the operator is setting using:
+
+```
+  --register-with-taints
+  --node-labels
+```
+
+This whitelist is passed as a command line flag to the apiserver. Admission
+control (either NodeRestriction or in a new admission control) will then prevent
+setting and modification by nodes of all taints and labels with keys not in the
+whitelist.
+
+Proposed flags to add to the apiserver to support this:
+
+```
+  --allow-node-self-labels
+  --allow-node-self-taints
+```
+
+### Removing self-delete from Node Permission
+
+Currently a node has permission to delete itself. A node will only delete itself
+when it's external name (inferred through the cloud provider) changes. This code
+path will never be executated on the majority of cloud providers and this
+capability undermines the usage of taints as a strong exclusion primitive.
+
+For example, suppose an operator sets a taint `compromised` on a node that they
+believe has been compromised. Currently, the compromised node could delete and
+recreate itself thereby removing the `compromised` taint.
+
+To prevent this, we will add a configuration option to the apiserver that
+indicates whether the NodeAuthorizer should allow self deletion.
```

Comment:
:+1:

Reply:
ExternalID has been marked as deprecated since 1.1

https://github.com/kubernetes/kubernetes/blob/release-1.1/pkg/api/v1/types.go#L1745-L1747

can we evaluate dropping it and the deletion cases associated with it altogether instead?

Reply:
Removing it would be preferable.